src/core/ext/filters/client_channel/lb_policy/xds/xds_override_host.cc

```diff
@@ -646,79 +619,58 @@ XdsOverrideHostLb::SubchannelWrapper::SubchannelWrapper(
     absl::optional<const std::string> key)
     : DelegatingSubchannel(std::move(subchannel)),
       key_(std::move(key)),
-      policy_(std::move(policy)) {}
+      policy_(std::move(policy)) {
+  auto watcher = std::make_unique<ConnectivityStateWatcher>();
+  watcher_ = watcher.get();
+  wrapped_subchannel()->WatchConnectivityState(std::move(watcher));
+}
 
 XdsOverrideHostLb::SubchannelWrapper::~SubchannelWrapper() {
+  wrapped_subchannel()->CancelConnectivityStateWatch(watcher_);
   if (key_.has_value()) {
     policy_->ResetSubchannel(*key_, this);
   }
 }
 
 void XdsOverrideHostLb::SubchannelWrapper::WatchConnectivityState(
     std::unique_ptr<ConnectivityStateWatcherInterface> watcher) {
-  auto watcher_id = watcher.get();
-  auto wrapper =
-      std::make_unique<ConnectivityStateWatcher>(std::move(watcher), Ref());
-  watchers_.emplace(watcher_id, wrapper.get());
-  wrapped_subchannel()->WatchConnectivityState(std::move(wrapper));
+  watcher_->watchers_.push_back(std::move(watcher));
 }
 
 void XdsOverrideHostLb::SubchannelWrapper::CancelConnectivityStateWatch(
     ConnectivityStateWatcherInterface* watcher) {
-  auto original_watcher = watchers_.find(watcher);
-  if (original_watcher != watchers_.end()) {
-    wrapped_subchannel()->CancelConnectivityStateWatch(
-        original_watcher->second);
-    watchers_.erase(original_watcher);
-  }
+  (void)std::remove_if(watcher_->watchers_.begin(), watcher_->watchers_.end(),
+                       [watcher](const auto& watcher_ptr) {
+                         return watcher_ptr.get() == watcher;
+                       });
 }
 
-//
-// factory
-//
-const JsonLoaderInterface* XdsOverrideHostLbConfig::JsonLoader(
-    const JsonArgs&) {
-  static const auto kJsonLoader =
-      JsonObjectLoader<XdsOverrideHostLbConfig>()
-          // Child policy config is parsed in JsonPostLoad
-          .Finish();
-  return kJsonLoader;
+grpc_pollset_set* XdsOverrideHostLb::SubchannelWrapper::
+    ConnectivityStateWatcher::interested_parties() {
+  if (interested_parties_ != nullptr) {
+    grpc_pollset_set_destroy(interested_parties_);
```

Comment:
I don't think this is safe.  The watcher will already have returned its interested_parties to the subchannel, and the subchannel may try to notify those interested_parties at any time.  We cannot delete the interested_parties out from under it.  Basically, the API contract here is that `interested_parties()` should return a single object that exists for the lifetime of the watcher.

I don't think it should be necessary to recreate the interested_parties set this way in the first place.